### PR TITLE
update jenkins api

### PR DESCRIPTION
**Pull request by soulseen**
Signed-off-by: soulseen <sunzhu@yunify.com>

### CODE REVIEW

**Comment by soulseen** [comment]
@zryfish @runzexia @wansir

File: pkg/apis/devops/v1alpha2/register.go [hunk 1]
```diff
@@ -271,27 +271,27 @@ func addWebService(c *restful.Container) error {
 		Param(webservice.QueryParameter("branch", "branch ").
 			Required(false).
 			DataFormat("branch=%s")).
-		Returns(http.StatusOK, RespOK, []devops.PipelineRun{}).
-		Writes([]devops.PipelineRun{}))
+		Returns(http.StatusOK, RespOK, []devops.PipeBranchRun{}).
+		Writes([]devops.PipeBranchRun{}))
```
Comment (zryfish) [inline pkg/apis/devops/v1alpha2/register.go]: `PipeBranch` seems like a new word, also not descriptive, I suggest use `PipelineBranch`

File: pkg/apis/devops/v1alpha2/register.go [hunk 2]
```diff
@@ -271,27 +271,27 @@ func addWebService(c *restful.Container) error {
...
 			DataFormat("start=%d")).
-		Returns(http.StatusOK, RespOK, devops.PipelineRun{}).
-		Writes(devops.PipelineRun{}))
+		Returns(http.StatusOK, RespOK, devops.PipeBranchRun{}).
+		Writes(devops.PipeBranchRun{}))
 
 	// match Jenkins api "/blue/rest/organizations/jenkins/pipelines/{projectName}/{pipelineName}/branches/{branchName}/runs/{runId}/nodes"
 	webservice.Route(webservice.GET("/devops/{projectName}/pipelines/{pipelineName}/branches/{branchName}/runs/{runId}/nodes").
-		To(devopsapi.GetPipelineRunNodes).
+		To(devopsapi.GetBranchPipeRunNodes).
```
Comment (zryfish) [inline pkg/apis/devops/v1alpha2/register.go]: what's the difference between `BranchPipe` and `PipeBranch`

File: pkg/apiserver/devops/devops.go [hunk 3]
```diff
@@ -57,37 +60,40 @@ func SearchPipelineRuns(req *restful.Request, resp *restful.Response) {
...
 
-	res, err := devops.GetPipelineRun(projectName, pipelineName, branchName, runId, req.Request)
+	res, err := devops.GetPipeBranchRun(projectName, pipelineName, branchName, runId, req.Request)
 	if err != nil {
 		parseErr(err, resp)
 		return
 	}
 
-	_ = resp.WriteAsJson(res)
+	resp.Header().Set(restful.HEADER_ContentType, "application/json")
```
Comment (zryfish) [inline pkg/apiserver/devops/devops.go]: not a good practice to directly hard coding `application/json` here, better to use a variable

File: pkg/models/devops/devops.go [hunk 4]
```diff
@@ -310,25 +296,124 @@ func GetCrumb(req *http.Request) (*Crumb, error) {
...
+		return nil, err
+	}
+
+	return res, err
+}
+
+func GetNodeSteps(projectName, pipelineName, branchName, runId, nodeId string, req *http.Request) ([]byte, error) {
+	baseUrl := fmt.Sprintf(JenkinsUrl+GetNodeStepsUrl+req.URL.RawQuery, projectName, pipelineName, branchName, runId, nodeId)
+	log.Infof("Jenkins-url: " + baseUrl)
+	//var res = new(NodeSteps)
```
Comment (zryfish) [inline pkg/models/devops/devops.go]: remove commented code

File: pkg/models/devops/json.go [hunk 5]
```diff
@@ -729,3 +730,329 @@ type Crumb struct {
 	Crumb             string `json:"crumb"`
 	CrumbRequestField string `json:"crumbRequestField"`
 }
+
+// CheckScriptCompile
```
Comment (zryfish) [inline pkg/models/devops/json.go]: for serializable struct, add `omitempty` to skip field generation to reduce response size

File: pkg/models/devops/devops.go [hunk 6]
```diff
@@ -310,25 +296,124 @@ func GetCrumb(req *http.Request) (*Crumb, error) {
...
+
+	return resBody, err
+}
+
+func CheckCron(req *http.Request) (*CheckCronRes, error) {
+	baseUrl := JenkinsUrl + CheckCronUrl + req.URL.RawQuery
+	log.Infof("Jenkins-url: " + baseUrl)
+	var res = new(CheckCronRes)
+
+	resp, err := http.Get(baseUrl)
```
Comment (runzexia) [inline pkg/models/devops/devops.go]: add basic auth,  only admin could access init-job

File: pkg/models/devops/devops.go [hunk 7]
```diff
@@ -310,25 +296,124 @@ func GetCrumb(req *http.Request) (*Crumb, error) {
...
 }
 
-// jenkins request and parse response
-func jenkinsClient(baseUrl string, req *http.Request, res interface{}) error {
-	resBody, err := Client(baseUrl, req)
+func CheckScriptCompile(req *http.Request) ([]byte, error) {
+	baseUrl := JenkinsUrl + CheckScriptCompileUrl
+	log.Infof("Jenkins-url: " + baseUrl)
+
+	resBody, err := sendJenkinsRequest(baseUrl, req)
```
Comment (runzexia) [inline pkg/models/devops/devops.go]: add basic auth, only admin could access init-job

File: pkg/apis/devops/v1alpha2/register.go (hunk 2 above)
Comment (soulseen) [reply pkg/apis/devops/v1alpha2/register.go]: `BranchPipe` mean all pipeline under the branch, and the `PipeBranch` mean all barnch under the pipeline.

File: pkg/apis/devops/v1alpha2/register.go (hunk 1 above)
Comment (soulseen) [reply pkg/apis/devops/v1alpha2/register.go]: I think it is more appropriate to use `BranchPipeline`.

File: pkg/apis/devops/v1alpha2/register.go (hunk 2 above)
Comment (zryfish) [reply pkg/apis/devops/v1alpha2/register.go]: ok, so what do you think of `GetPipelinesByBranch` `GetBranchesByPipeline`

Comment (soulseen) [reply pkg/apis/devops/v1alpha2/register.go]: ok

File: pkg/models/devops/devops.go (hunk 6 above)
Comment (soulseen) [reply pkg/models/devops/devops.go]: check cron and check pipeline script only admin?

Comment (runzexia) [reply pkg/models/devops/devops.go]: use admin  account in apiserver, but all user could use api

Comment (soulseen) [reply pkg/models/devops/devops.go]: ok

File: pkg/apiserver/devops/devops.go (hunk 3 above)
Comment (soulseen) [reply pkg/apiserver/devops/devops.go]: response ok.

File: pkg/models/devops/devops.go [hunk 8]
```diff
@@ -19,29 +19,54 @@ package devops
...
 	"io/ioutil"
+	"kubesphere.io/kubesphere/pkg/simple/client/admin_jenkins"
 	"net/http"
 	"net/url"
+	"strings"
 	"time"
 )
 
-var JenkinsUrl string
+const MIME_FORM = "application/x-www-form-urlencoded"
```
Comment (runzexia) [inline pkg/models/devops/devops.go]: how about move to `constants`  package

File: pkg/apis/devops/v1alpha2/register.go [hunk 9]
```diff
@@ -517,6 +517,130 @@ func addWebService(c *restful.Container) error {
...
+	// match /job/init-job/descriptorByName/org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition/checkScriptCompile
+	webservice.Route(webservice.POST("/devops/check/scriptcompile").
+		To(devopsapi.CheckScriptCompile).
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Consumes("application/x-www-form-urlencoded", "charset=utf-8").
+		Produces("application/json", "charset=utf-8").
+		Doc("Check pipeline script compile.").
+		Reads("value=\"\"").
+		Returns(http.StatusOK, RespOK, devops.CheckScript{}).
+		Writes(devops.CheckScript{}))
```
Comment (runzexia) [inline pkg/apis/devops/v1alpha2/register.go]: I recommend returning an array, although it is a single error most of the time.

File: pkg/simple/client/admin_jenkins/jenkins.go [hunk 10]
```diff
@@ -40,6 +40,10 @@ func init() {
 	flag.IntVar(&jenkinsMaxConn, "jenkins-max-conn", 20, "max conn to jenkins")
 }
 
+func GetJenkinsFlag() (string,string,string){
```
Comment (runzexia) [inline pkg/simple/client/admin_jenkins/jenkins.go]: how about return structs?

File: cmd/ks-apiserver/app/server.go [hunk 11]
```diff
@@ -98,6 +99,7 @@ func Run(s *options.ServerRunOptions) error {
 }
 
 func initializeAdminJenkins() {
+	devops.PreCheckJenkins()
```
Comment (runzexia) [inline cmd/ks-apiserver/app/server.go]: admin_jenkins.Client also check login

File: pkg/models/devops/devops.go [hunk 12]
```diff
@@ -19,29 +19,54 @@ package devops
...
 
-func init() {
-	flag.StringVar(&JenkinsUrl, "jenkins-url", "http://ks-jenkins.kubesphere-devops-system.svc.cluster.local:80", "jenkins server host")
+var (
```
Comment (runzexia) [inline pkg/models/devops/devops.go]: return the struct in `admin_jenkins`, remove this part of the var

File: cmd/ks-apiserver/app/server.go (hunk 11 above)
Comment (runzexia) [reply cmd/ks-apiserver/app/server.go]: Print more logs in the client method might be a better option

File: pkg/apis/devops/v1alpha2/register.go (hunk 9 above)
Comment (soulseen) [reply pkg/apis/devops/v1alpha2/register.go]: yes, returning an array would be better if the result is failure, but if the result is successful, jenkins will return a map, not array. So for structural unity, i think return a struct will be better

File: cmd/ks-apiserver/app/server.go (hunk 11 above)
Comment (soulseen) [reply cmd/ks-apiserver/app/server.go]: this check is repeated, i remove it and go to `admin_jenkins.Client`.

**Pull request merged**